### ISSUE TITLE
Cannot bootstrap from docker behind proxy - required envs are being sanitized.

### ISSUE THREAD

**Issue by dtrudg** [apptainer/singularity#764]
## Version of Singularity:

2.3

### Expected behavior

When an http/https proxy must be used, singularity should be able to bootstrap a definition that is importing from docker hub / other registry.

### Actual behavior

Proxy environment variables do not reach the python urllib(2) code that is used to connect to the docker registry, so the connection fails and it is not possible to bootstrap from docker behind a proxy.

Def file..
```
BootStrap: docker
From: biocontainers/abyss
IncludeCmd: yes
```

Command..
```
sudo http_proxy=http://proxy.swmed.edu:3128 /cm/shared/apps/singularity/2.3/bin/singularity bootstrap tmp/ Singularity
```

Processes spawned..
```
root     12413  1.7  0.0 178008  2712 pts/5    S+   10:34   0:00 sudo HTTPS_PROXY=http://proxy.swmed.edu:3128 /cm/shared/apps/singularity/2.3/bin/singularity bootstrap tmp/ Singularity
root     12424  3.2  0.0  11332  1300 pts/5    S+   10:34   0:00 /bin/bash /cm/shared/apps/singularity/2.3/libexec/singularity/bootstrap-scripts/main-deffile.sh
root     12464  0.6  0.0  11332  1300 pts/5    S+   10:34   0:00 /bin/bash /cm/shared/apps/singularity/2.3/libexec/singularity/bootstrap-scripts/deffile-driver-docker.sh
root     12472 18.6  0.0 107020  9508 pts/5    S+   10:34   0:00 python /cm/shared/apps/singularity/2.3/libexec/singularity/python/import.py
```

Checking environments for the deeper processes, we see the proxy environment var is not available:

```
dtrudgian@Nucleus006:/proc/11829
10:34 AM $ sudo grep proxy /proc/12424/environ
dtrudgian@Nucleus006:/proc/11829
10:34 AM $ sudo grep proxy /proc/12464/environ
dtrudgian@Nucleus006:/proc/11829
10:34 AM $ sudo grep proxy /proc/12472/environ
```

Therefore, proxy is not used and the import fails:

```
Sanitizing environment
Building from bootstrap definition recipe
Adding base Singularity environment to container
Traceback (most recent call last):
  File "/cm/shared/apps/singularity/2.3/libexec/singularity/python/import.py", line 107, in <module>
    main()
  File "/cm/shared/apps/singularity/2.3/libexec/singularity/python/import.py", line 83, in main
    layerfile=LAYERFILE)
  File "/cm/shared/apps/singularity/2.3/libexec/singularity/python/docker/main.py", line 82, in IMPORT
    client = DockerApiConnection(image=image,auth=auth)
  File "/cm/shared/apps/singularity/2.3/libexec/singularity/python/docker/api.py", line 83, in __init__
    self.load_image(kwargs['image'])
  File "/cm/shared/apps/singularity/2.3/libexec/singularity/python/docker/api.py", line 120, in load_image
    self.update_token()
  File "/cm/shared/apps/singularity/2.3/libexec/singularity/python/docker/api.py", line 131, in update_token
    response = self.get_tags(return_response=True)
  File "/cm/shared/apps/singularity/2.3/libexec/singularity/python/docker/api.py", line 217, in get_tags
    return_response=return_response)
  File "/cm/shared/apps/singularity/2.3/libexec/singularity/python/base.py", line 247, in get
    return_response=return_response)
  File "/cm/shared/apps/singularity/2.3/libexec/singularity/python/base.py", line 264, in submit_request
    response = urlopen(request)
  File "/usr/lib64/python2.6/urllib2.py", line 126, in urlopen
    return _opener.open(url, data, timeout)
  File "/usr/lib64/python2.6/urllib2.py", line 391, in open
    response = self._open(req, data)
  File "/usr/lib64/python2.6/urllib2.py", line 409, in _open
    '_open', req)
  File "/usr/lib64/python2.6/urllib2.py", line 369, in _call_chain
    result = func(*args)
  File "/usr/lib64/python2.6/urllib2.py", line 1198, in https_open
    return self.do_open(httplib.HTTPSConnection, req)
  File "/usr/lib64/python2.6/urllib2.py", line 1165, in do_open
    raise URLError(err)
urllib2.URLError: <urlopen error [Errno 110] Connection timed out>
```

Can http_proxy, https_proxy be passed through during the environment sanitization?

**Comment by dtrudg** [comment]
NB - I can workaround by doing an unpriveleged `singularity pull` (which uses proxy OK) and then scripting changes to the resulting container - but to be able to do a one step bootstrap would be nice.

**Comment by dtrudg** [comment]
I guess a fix would be similar to #761 assuming it's considered safe to pass through the http_proxy and https_proxy envs?

**Comment by vsoch** [comment]
We could try passing the proxy through to the python... maybe as an environment variable? @gmkurter is this safe to try?

**Comment by dtrudg** [comment]
NB - current hacky workaround is to add...

```python
os.environ['http_proxy'] = "http://host:port"
os.environ['https_proxy'] = "http://host:port"
```
... into `libexec/python/base.py` before the urllib import lines.

**Comment by jekriske-lilly** [comment]
Thanks for the hack, this is still a problem in 2.4

**Comment by dtrudg** [comment]
Closing this as it has been fixed in #1151.
